### PR TITLE
store-gateway: fix duration metrics

### PR DESCRIPTION
**Pull request by dimitarvdimitrov**
* don't double count streamingSeriesOtherDuration and streamingSeriesFetchSeriesAndChunksDuration when using querier<>store-gateway streaming
* replace streamingSeriesFetchSeriesAndChunksDuration with streamingSeriesWaitBatchLoadedDuration and only record it once
* remove cortex_bucket_store_series_refs_fetch_duration_seconds because it is tracking the time per block, not per request and isn't directly useful in diagnosing slow store-gateway requests; it is also not used in dashboards
* record send/encode duration for sending stats
* record send/encode duration even if sending/encoding fails
* calculate "other" time based on the total time of the request (for example now it includes reading the bucket index, waiting for the query gate)

### CODE REVIEW

File: pkg/storegateway/stats.go [hunk 1]
```diff
@@ -70,12 +67,13 @@ type queryStats struct {
...
-	streamingSeriesEncodeResponseDuration       time.Duration
-	streamingSeriesSendResponseDuration         time.Duration
-	streamingSeriesOtherDuration                time.Duration
-	streamingSeriesIndexHeaderLoadDuration      time.Duration
+	streamingSeriesExpandPostingsDuration  time.Duration
+	streamingSeriesEncodeResponseDuration  time.Duration
+	streamingSeriesSendResponseDuration    time.Duration
+	streamingSeriesIndexHeaderLoadDuration time.Duration
+
+	// streamingSeriesAmbientTime is the total WAL time spent serving the request. It includes all other durations.
```
Comment (56quarters) [inline pkg/storegateway/stats.go:75]: Is this supposed to be `wall`, not `WAL`?
```suggestion
	// streamingSeriesAmbientTime is the total wall clock time spent serving the request. It includes all other durations.
```

**Review by 56quarters — APPROVED** [review]
LGTM

File: pkg/storegateway/bucket.go [hunk 2]
```diff
@@ -1237,12 +1226,16 @@ func (s *BucketStore) recordStreamingSeriesStats(stats *queryStats) {
...
 
-	s.metrics.streamingSeriesRefsFetchDuration.Observe(stats.streamingSeriesFetchRefsDuration.Seconds())
-
 	s.metrics.streamingSeriesRequestDurationByStage.WithLabelValues("expand_postings").Observe(stats.streamingSeriesExpandPostingsDuration.Seconds())
-	s.metrics.streamingSeriesRequestDurationByStage.WithLabelValues("fetch_series_and_chunks").Observe(stats.streamingSeriesFetchSeriesAndChunksDuration.Seconds())
-	s.metrics.streamingSeriesRequestDurationByStage.WithLabelValues("other").Observe(stats.streamingSeriesOtherDuration.Seconds())
+	s.metrics.streamingSeriesRequestDurationByStage.WithLabelValues("fetch_series_and_chunks").Observe(stats.streamingSeriesBatchLoadDuration.Seconds())
 	s.metrics.streamingSeriesRequestDurationByStage.WithLabelValues("load_index_header").Observe(stats.streamingSeriesIndexHeaderLoadDuration.Seconds())
+
+	specializedTime := stats.streamingSeriesExpandPostingsDuration +
```
[the remark this answers is not in the document]
Comment (dimitarvdimitrov) [reply pkg/storegateway/bucket.go:1233]: how about `trackedTime`? It's the time we have recorded and tracked specifically. Anything else in the ambient time will be categorized as `other`. Alternatively, `knownTime`?

[the remark this answers is not in the document]
Comment (dimitarvdimitrov) [reply pkg/storegateway/bucket.go]: i went with `categorizedTime`

**Pull request merged**